### PR TITLE
impl in out components filter

### PR DESCRIPTION
**Pull request by shivamka1**
Fixes https://github.com/Pometry/Raphtory/issues/2264

### CODE REVIEW

File: python/python/raphtory/algorithms/__init__.pyi [hunk 1]
```diff
@@ -438,7 +438,7 @@ def strongly_connected_components(graph: GraphView) -> NodeStateUsize:
         NodeStateUsize: Mapping of nodes to their component ids
     """
 
-def in_components(graph: GraphView) -> NodeStateNodes:
+def in_components(graph: GraphView, filter=None) -> NodeStateNodes:
```
Comment (ljeub-pometry) [inline python/python/raphtory/algorithms/__init__.pyi]: filter needs to be in the docstring

File: python/python/raphtory/algorithms/__init__.pyi [hunk 2]
```diff
@@ -449,7 +449,7 @@ def in_components(graph: GraphView) -> NodeStateNodes:
         NodeStateNodes: Mapping of nodes to the nodes in their 'in-component'
     """
 
-def in_component(node: Node) -> NodeStateUsize:
+def in_component(node: Node, filter=None) -> NodeStateUsize:
```
Comment (ljeub-pometry) [inline python/python/raphtory/algorithms/__init__.pyi]: filter needs to be in the docstring

File: python/python/raphtory/algorithms/__init__.pyi [hunk 3]
```diff
@@ -460,7 +460,7 @@ def in_component(node: Node) -> NodeStateUsize:
        NodeStateUsize: Mapping of nodes in the in-component to the distance from the starting node.
     """
 
-def out_components(graph: GraphView) -> NodeStateNodes:
+def out_components(graph: GraphView, filter=None) -> NodeStateNodes:
```
Comment (ljeub-pometry) [inline python/python/raphtory/algorithms/__init__.pyi]: filter needs to be in the docstring

File: python/python/raphtory/algorithms/__init__.pyi [hunk 4]
```diff
@@ -471,7 +471,7 @@ def out_components(graph: GraphView) -> NodeStateNodes:
         NodeStateNodes: Mapping of nodes to the nodes within their 'out-component'
     """
 
-def out_component(node: Node) -> NodeStateUsize:
+def out_component(node: Node, filter=None) -> NodeStateUsize:
```
Comment (ljeub-pometry) [inline python/python/raphtory/algorithms/__init__.pyi]: filter needs to be in the docstring

File: raphtory/src/algorithms/components/in_components.rs [hunk 5]
```diff
@@ -39,7 +44,38 @@ pub fn in_components<G>(g: &G, threads: Option<usize>) -> NodeState<'static, Nod
 where
     G: StaticGraphViewOps,
 {
-    let ctx: Context<G, ComputeStateVec> = g.into();
+    in_components_filtered(g, threads, Unfiltered).expect("Unfiltered should never fail")
+}
+
+/// Computes the in components of each node in the graph
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/in_components.rs]: Add the updated docstring to the original function as well.

File: raphtory/src/algorithms/components/in_components.rs [hunk 6]
```diff
@@ -39,7 +44,38 @@ pub fn in_components<G>(g: &G, threads: Option<usize>) -> NodeState<'static, Nod
...
+///
+/// # Arguments
+///
+/// - `g` - A reference to the graph
+/// - `threads` - Number of threads to use
+/// - `filter` - Filter
+///
+/// # Returns
+///
+/// An [AlgorithmResult] containing the filtered mapping from each node to a vector of node ids (the nodes in component)
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/in_components.rs]: `AlgorithmResult` doesn't exist anymore for a long time, also this description doesn't match the actual output

Comment (ljeub-pometry) [reply raphtory/src/algorithms/components/in_components.rs]: maybe get rid of the Returns part

File: raphtory/src/algorithms/components/in_components.rs (hunk 5 above)
Comment (ljeub-pometry) [reply raphtory/src/algorithms/components/in_components.rs]: mention here that the in-component is computed over the filtered graph

File: raphtory/src/algorithms/components/in_components.rs [hunk 7]
```diff
@@ -39,7 +44,38 @@ pub fn in_components<G>(g: &G, threads: Option<usize>) -> NodeState<'static, Nod
...
+    g: &G,
+    threads: Option<usize>,
+    filter: F,
+) -> Result<NodeState<'static, Nodes<'static, G>, G>, GraphError>
+where
+    G: StaticGraphViewOps,
+    F: CreateFilter + 'static,
+    F::EntityFiltered<'static, F::FilteredGraph<'static, G>>: StaticGraphViewOps,
+{
+    type FG<G, F> = <F as CreateFilter>::EntityFiltered<
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/in_components.rs]: the `FG` type doesn't need the generics, just `type FG = ...` should work

File: raphtory/src/algorithms/components/in_components.rs [hunk 8]
```diff
@@ -98,19 +134,44 @@ where
...
 ///
-pub fn in_component<'graph, G: GraphViewOps<'graph>>(
+pub fn in_component<'graph, G>(node: NodeView<'graph, G>) -> NodeState<'graph, usize, G>
+where
+    G: GraphViewOps<'graph>,
+{
+    in_component_filtered(node, Unfiltered).expect("Unfiltered should never fail")
+}
+
+/// Computes the in-component of a given node in the graph
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/in_components.rs]: same comment as for the `in-components`, mention this is computed over the filtered graph

File: raphtory/src/algorithms/components/out_components.rs [hunk 9]
```diff
@@ -39,7 +44,39 @@ pub fn out_components<G>(g: &G, threads: Option<usize>) -> NodeState<'static, No
 where
     G: StaticGraphViewOps,
 {
-    let ctx: Context<G, ComputeStateVec> = g.into();
+    out_components_filtered(g, threads, Unfiltered).expect("Unfiltered should never fail")
+}
+
+/// Computes the out components of each node in the graph
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/out_components.rs]: Same docstring tweaks as for in-components

File: raphtory/src/algorithms/components/out_components.rs [hunk 10]
```diff
@@ -39,7 +44,39 @@ pub fn out_components<G>(g: &G, threads: Option<usize>) -> NodeState<'static, No
...
+    g: &G,
+    threads: Option<usize>,
+    filter: F,
+) -> Result<NodeState<'static, Nodes<'static, G>, G>, GraphError>
+where
+    G: StaticGraphViewOps,
+    F: CreateFilter + 'static,
+    F::EntityFiltered<'static, F::FilteredGraph<'static, G>>: StaticGraphViewOps,
+{
+    type FG<G, F> = <F as CreateFilter>::EntityFiltered<
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/out_components.rs]: no need for generic parameters on `FG`

File: raphtory/src/algorithms/components/out_components.rs [hunk 11]
```diff
@@ -39,7 +44,39 @@ pub fn out_components<G>(g: &G, threads: Option<usize>) -> NodeState<'static, No
...
+        'static,
+        <F as CreateFilter>::FilteredGraph<'static, G>,
+    >;
+    let filtered = g.filter(filter)?;
+    let ctx: Context<FG<G, F>, ComputeStateVec> = (&filtered).into();
+
+    // let ctx: Context<G, ComputeStateVec> = g.into();
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/out_components.rs]: delete comment

File: raphtory/src/python/packages/algorithms.rs [hunk 12]
```diff
@@ -174,11 +175,15 @@ pub fn connected_components(graph: &PyDiskGraph) -> Vec<usize> {
 /// Returns:
 ///     NodeStateNodes: Mapping of nodes to the nodes in their 'in-component'
 #[pyfunction]
-#[pyo3(signature = (graph))]
+#[pyo3(signature = (graph, filter = None))]
```
Comment (ljeub-pometry) [inline raphtory/src/python/packages/algorithms.rs:179]: update the docstring to mention `filter`

File: raphtory/src/python/packages/algorithms.rs [hunk 13]
```diff
@@ -189,9 +194,15 @@ pub fn in_components(
 /// Returns:
 ///    NodeStateUsize: Mapping of nodes in the in-component to the distance from the starting node.
 #[pyfunction]
-#[pyo3(signature = (node))]
-pub fn in_component(node: &PyNode) -> NodeState<'static, usize, DynamicGraph> {
-    components::in_component(node.node.clone())
+#[pyo3(signature = (node, filter = None))]
```
Comment (ljeub-pometry) [inline raphtory/src/python/packages/algorithms.rs:199]: update the docstring to mention `filter`

File: raphtory/src/python/packages/algorithms.rs [hunk 14]
```diff
@@ -202,11 +213,15 @@ pub fn in_component(node: &PyNode) -> NodeState<'static, usize, DynamicGraph> {
 /// Returns:
 ///     NodeStateNodes: Mapping of nodes to the nodes within their 'out-component'
 #[pyfunction]
-#[pyo3(signature = (graph))]
+#[pyo3(signature = (graph, filter = None))]
 pub fn out_components(
     graph: &PyGraphView,
-) -> NodeState<'static, Nodes<'static, DynamicGraph>, DynamicGraph> {
-    components::out_components(&graph.graph, None)
+    filter: Option<PyFilterExpr>,
```
Comment (ljeub-pometry) [inline raphtory/src/python/packages/algorithms.rs:222]: mention filter in the docs

File: raphtory/src/python/packages/algorithms.rs [hunk 15]
```diff
@@ -217,9 +232,15 @@ pub fn out_components(
...
 ///    NodeStateUsize: A NodeState mapping the nodes in the out-component to their distance from the starting node.
 #[pyfunction]
-#[pyo3(signature = (node))]
-pub fn out_component(node: &PyNode) -> NodeState<'static, usize, DynamicGraph> {
-    components::out_component(node.node.clone())
+#[pyo3(signature = (node, filter = None))]
+pub fn out_component(
+    node: &PyNode,
+    filter: Option<PyFilterExpr>,
+) -> Result<NodeState<'static, usize, DynamicGraph>, GraphError> {
```
Comment (ljeub-pometry) [inline raphtory/src/python/packages/algorithms.rs:243]: mention filter in the docs

File: raphtory/tests/algo_tests/components.rs [hunk 16]
```diff
@@ -301,13 +326,100 @@ mod in_component_test {
...
+
+        graph.add_edge(1, 1, 2, NO_PROPS, Some("A")).unwrap();
+        graph.add_edge(1, 2, 4, NO_PROPS, Some("A")).unwrap();
+        graph.add_edge(1, 4, 6, NO_PROPS, Some("A")).unwrap();
+
+        graph.add_edge(1, 2, 5, NO_PROPS, Some("B")).unwrap();
+        graph.add_edge(1, 5, 4, NO_PROPS, Some("B")).unwrap();
+
+        // Sanity: with A-only filter, 5 should disappear
+        let filter = GraphFilter.layer("A");
```
Comment (ljeub-pometry) [inline raphtory/tests/algo_tests/components.rs:346]: can you add a check for testing that the filter no longer applies to the output nodes

Comment (ljeub-pometry) [reply raphtory/tests/algo_tests/components.rs:346]: i.e., double-check that the filter acts like `select` and not like `filter`

File: raphtory/tests/algo_tests/components.rs [hunk 17]
```diff
@@ -301,13 +326,100 @@ mod in_component_test {
             assert_eq!(map, correct);
         });
     }
+
```
Comment (ljeub-pometry) [inline raphtory/tests/algo_tests/components.rs:333]: We should also test this with node filters and edge filters

File: raphtory/src/algorithms/components/in_components.rs [hunk 18]
```diff
@@ -101,7 +89,7 @@ where
         Step::Done
     });
 
-    let mut runner: TaskRunner<FG<G, F>, ComputeStateVec> = TaskRunner::new(ctx);
+    let mut runner: TaskRunner<_, _> = TaskRunner::new(ctx);
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/in_components.rs]: might as well remove the entire type annotation

File: raphtory/src/algorithms/components/out_components.rs [hunk 19]
```diff
@@ -102,7 +89,7 @@ where
         Step::Done
     });
 
-    let mut runner: TaskRunner<FG<G, F>, _> = TaskRunner::new(ctx);
+    let mut runner: TaskRunner<_, _> = TaskRunner::new(ctx);
```
Comment (ljeub-pometry) [inline raphtory/src/algorithms/components/out_components.rs]: type annotation not needed

File: raphtory/src/graphgen/erdos_renyi.rs [hunk 20]
```diff
@@ -13,7 +13,7 @@ use crate::{
         graph::graph::Graph,
     },
     errors::GraphError,
-    prelude::NO_PROPS,
+    prelude::{NodeStateOps, NO_PROPS},
```
Comment (ljeub-pometry) [inline raphtory/src/graphgen/erdos_renyi.rs]: this import should be in the test module

File: raphtory/src/python/packages/algorithms.rs [hunk 21]
```diff
@@ -171,6 +171,7 @@ pub fn connected_components(graph: &PyDiskGraph) -> Vec<usize> {
 ///
 /// Arguments:
 ///     graph (GraphView): Raphtory graph
+///     filter (Option<Filter>): Optional filter
```
Comment (ljeub-pometry) [inline raphtory/src/python/packages/algorithms.rs]: this should be `filter (Filter, optional): Optional filter`

File: raphtory/tests/algo_tests/components.rs [hunk 22]
```diff
@@ -405,7 +409,60 @@ mod in_component_test {
...
+    #[test]
+    fn in_component_filtered_returns_nodes_that_are_unfiltered_for_future_traversals() {
+        let graph = Graph::new();
+
+        graph.add_edge(1, 2, 4, NO_PROPS, Some("A")).unwrap();
+        graph.add_edge(1, 99, 2, NO_PROPS, Some("B")).unwrap();
+        graph.add_edge(1, 100, 99, NO_PROPS, Some("B")).unwrap();
+
+        let mut unfiltered_ids: Vec<u64> =
+            in_component_filtered(graph.node(4).unwrap(), GraphFilter.layer("A"))
```
Comment (ljeub-pometry) [inline raphtory/tests/algo_tests/components.rs:428]: should add an assert that this is indeed returning only 2

File: raphtory/tests/algo_tests/components.rs [hunk 23]
```diff
@@ -405,7 +409,60 @@ mod in_component_test {
...
+                .unwrap()
+                .nodes()
+                .in_neighbours()
+                .iter()
+                .flat_map(|ns| ns.iter().filter_map(|c| c.id().as_u64()))
+                .collect();
+
+        unfiltered_ids.sort();
+        unfiltered_ids.dedup();
```
Comment (ljeub-pometry) [inline raphtory/tests/algo_tests/components.rs]: get rid of these as there is only one element

File: raphtory/tests/algo_tests/components.rs [hunk 24]
```diff
@@ -626,6 +686,58 @@ mod components_test {
...
+            out_component_filtered(graph.node(1).unwrap(), GraphFilter.layer("A"))
+                .unwrap()
+                .nodes()
+                .out_neighbours()
+                .iter()
+                .flat_map(|ns| ns.iter().filter_map(|c| c.id().as_u64()))
+                .collect();
+
+        unfiltered_ids.sort();
+        unfiltered_ids.dedup();
```
Comment (ljeub-pometry) [inline raphtory/tests/algo_tests/components.rs]: remove these (especially the dedup) as there should only be one element
